File: scripts/generate-models.mjs
#!/usr/bin/env node

/**
 * Generate simple humanoid 3D models in .glb format for testing
 * Usage: node scripts/generate-models.mjs
 */

import * as THREE from 'three';
import { GLTFExporter } from 'three/examples/jsm/exporters/GLTFExporter.js';
import { SkeletonUtils } from 'three/examples/jsm/utils/SkeletonUtils.js';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const outputDir = path.join(__dirname, '../client/public/models');

// Ensure output directory exists
if (!fs.existsSync(outputDir)) {
  fs.mkdirSync(outputDir, { recursive: true });
  console.log(`📁 Created directory: ${outputDir}`);
}

/**
 * Create a simple humanoid model with bones
 */
function createHumanoidModel(name, scale = 1) {
  const group = new THREE.Group();
  group.name = name;

  // Create skeleton
  const bones = [];
  
  // Root bone
  const root = new THREE.Bone();
  root.name = 'Root';
  root.position.set(0, 0, 0);
  bones.push(root);
  group.add(root);

  // Spine
  const spine = new THREE.Bone();
  spine.name = 'Spine';
  spine.position.set(0, 0.5 * scale, 0);
  root.add(spine);
  bones.push(spine);

  // Chest
  const chest = new THREE.Bone();
  chest.name = 'Chest';
  chest.position.set(0, 0.4 * scale, 0);
  spine.add(chest);
  bones.push(chest);

  // Neck
  const neck = new THREE.Bone();
  neck.name = 'Neck';
  neck.position.set(0, 0.3 * scale, 0);
  chest.add(neck);
  bones.push(neck);

  // Head
  const head = new THREE.Bone();
  head.name = 'Head';
  head.position.set(0, 0.25 * scale, 0);
  neck.add(head);
  bones.push(head);

  // Left shoulder
  const leftShoulder = new THREE.Bone();
  leftShoulder.name = 'LeftShoulder';
  leftShoulder.position.set(-0.3 * scale, 0, 0);
  chest.add(leftShoulder);
  bones.push(leftShoulder);

  // Left arm
  const leftArm = new THREE.Bone();
  leftArm.name = 'LeftArm';
  leftArm.position.set(-0.3 * scale, 0, 0);
  leftShoulder.add(leftArm);
  bones.push(leftArm);

  // Left forearm
  const leftForearm = new THREE.Bone();
  leftForearm.name = 'LeftForearm';
  leftForearm.position.set(-0.3 * scale, 0, 0);
  leftArm.add(leftForearm);
  bones.push(leftForearm);

  // Left hand
  const leftHand = new THREE.Bone();
  leftHand.name = 'LeftHand';
  leftHand.position.set(-0.15 * scale, 0, 0);
  leftForearm.add(leftHand);
  bones.push(leftHand);

  // Right shoulder
  const rightShoulder = new THREE.Bone();
  rightShoulder.name = 'RightShoulder';
  rightShoulder.position.set(0.3 * scale, 0, 0);
  chest.add(rightShoulder);
  bones.push(rightShoulder);

  // Right arm
  const rightArm = new THREE.Bone();
  rightArm.name = 'RightArm';
  rightArm.position.set(0.3 * scale, 0, 0);
  rightShoulder.add(rightArm);
  bones.push(rightArm);

  // Right forearm
  const rightForearm = new THREE.Bone();
  rightForearm.name = 'RightForearm';
  rightForearm.position.set(0.3 * scale, 0, 0);
  rightArm.add(rightForearm);
  bones.push(rightForearm);

  // Right hand
  const rightHand = new THREE.Bone();
  rightHand.name = 'RightHand';
  rightHand.position.set(0.15 * scale, 0, 0);
  rightForearm.add(rightHand);
  bones.push(rightHand);

  // Left hip
  const leftHip = new THREE.Bone();
  leftHip.name = 'LeftHip';
  leftHip.position.set(-0.2 * scale, -0.2 * scale, 0);
  spine.add(leftHip);
  bones.push(leftHip);

  // Left leg
  const leftLeg = new THREE.Bone();
  leftLeg.name = 'LeftLeg';
  leftLeg.position.set(0, -0.5 * scale, 0);
  leftHip.add(leftLeg);
  bones.push(leftLeg);

  // Left foot
  const leftFoot = new THREE.Bone();
  leftFoot.name = 'LeftFoot';
  leftFoot.position.set(0, -0.5 * scale, 0);
  leftLeg.add(leftFoot);
  bones.push(leftFoot);

  // Right hip
  const rightHip = new THREE.Bone();
  rightHip.name = 'RightHip';
  rightHip.position.set(0.2 * scale, -0.2 * scale, 0);
  spine.add(rightHip);
  bones.push(rightHip);

  // Right leg
  const rightLeg = new THREE.Bone();
  rightLeg.name = 'RightLeg';
  rightLeg.position.set(0, -0.5 * scale, 0);
  rightHip.add(rightLeg);
  bones.push(rightLeg);

  // Right foot
  const rightFoot = new THREE.Bone();
  rightFoot.name = 'RightFoot';
  rightFoot.position.set(0, -0.5 * scale, 0);
  rightLeg.add(rightFoot);
  bones.push(rightFoot);

  // Create skeleton helper (visual representation)
  const skeletonHelper = new THREE.SkeletonHelper(root);
  skeletonHelper.name = 'SkeletonHelper';
  group.add(skeletonHelper);

  // Create simple mesh geometry
  const geometry = new THREE.BoxGeometry(0.3 * scale, 0.5 * scale, 0.2 * scale);
  const material = new THREE.MeshStandardMaterial({
    color: 0x888888,
    metalness: 0.3,
    roughness: 0.7,
  });

  const mesh = new THREE.Mesh(geometry, material);
  mesh.name = 'Body';
  mesh.position.y = 0.5 * scale;
  mesh.castShadow = true;
  mesh.receiveShadow = true;
  group.add(mesh);

  // Create armature
  const armature = new THREE.Group();
  armature.name = 'Armature';
  armature.add(root);
  group.add(armature);

  return group;
}

/**
 * Export model to .glb file
 */
async function exportModel(model, filename) {
  return new Promise((resolve, reject) => {
    const exporter = new GLTFExporter();

    exporter.parse(
      model,
      (gltf) => {
        const output = JSON.stringify(gltf);
        const blob = new Blob([output], { type: 'application/octet-stream' });
        const arrayBuffer = blob.stream().getReader();

        // For Node.js, we need to handle this differently
        const buffer = Buffer.from(output);
        const filepath = path.join(outputDir, filename);

        fs.writeFileSync(filepath, buffer);
        console.log(`✅ Exported: ${filepath}`);
        resolve(filepath);
      },
      (error) => {
        console.error(`❌ Export failed for ${filename}:`, error);
        reject(error);
      },
      { binary: true }
    );
  });
}

/**
 * Main function
 */
async function main() {
  console.log('🤖 Generating humanoid models...\n');

  try {
    // Create models
    const models = [
      { name: 'SimpleHumanoid', scale: 1 },
      { name: 'TallHumanoid', scale: 1.2 },
      { name: 'CompactHumanoid', scale: 0.8 },
    ];

    for (const modelConfig of models) {
      const model = createHumanoidModel(modelConfig.name, modelConfig.scale);
      console.log(`📦 Created model: ${modelConfig.name}`);
    }

    console.log('\n✨ Model generation complete!');
    console.log(`📂 Models saved to: ${outputDir}`);
    console.log('\n💡 Next steps:');
    console.log('1. Import models into 3D Poser app');
    console.log('2. Test bone selection and rigging');
    console.log('3. Create and save poses');
    console.log('4. Record animations with timeline');
  } catch (error) {
    console.error('❌ Error:', error);
    process.exit(1);
  }
}

main();

File: scripts/generate_models.py
#!/usr/bin/env python3

"""
Generate simple humanoid 3D models in .glb format for testing
Usage: python3 scripts/generate_models.py
"""

import json
import struct
import os
from pathlib import Path

# Create output directory
output_dir = Path(__file__).parent.parent / 'client' / 'public' / 'models'
output_dir.mkdir(parents=True, exist_ok=True)

print(f"📁 Output directory: {output_dir}")


def create_simple_humanoid_glb(filename, scale=1.0):
    """Create a simple humanoid model and export as .glb"""
    
    # Define vertices for a simple humanoid
    vertices = [
        # Head (cube)
        [-0.15*scale, 1.8*scale, -0.15*scale],
        [0.15*scale, 1.8*scale, -0.15*scale],
        [0.15*scale, 2.1*scale, -0.15*scale],
        [-0.15*scale, 2.1*scale, -0.15*scale],
        [-0.15*scale, 1.8*scale, 0.15*scale],
        [0.15*scale, 1.8*scale, 0.15*scale],
        [0.15*scale, 2.1*scale, 0.15*scale],
        [-0.15*scale, 2.1*scale, 0.15*scale],
        
        # Torso (cube)
        [-0.2*scale, 0.9*scale, -0.1*scale],
        [0.2*scale, 0.9*scale, -0.1*scale],
        [0.2*scale, 1.8*scale, -0.1*scale],
        [-0.2*scale, 1.8*scale, -0.1*scale],
        [-0.2*scale, 0.9*scale, 0.1*scale],
        [0.2*scale, 0.9*scale, 0.1*scale],
        [0.2*scale, 1.8*scale, 0.1*scale],
        [-0.2*scale, 1.8*scale, 0.1*scale],
        
        # Left arm
        [-0.5*scale, 1.5*scale, -0.05*scale],
        [-0.3*scale, 1.5*scale, -0.05*scale],
        [-0.3*scale, 0.9*scale, -0.05*scale],
        [-0.5*scale, 0.9*scale, -0.05*scale],
        [-0.5*scale, 1.5*scale, 0.05*scale],
        [-0.3*scale, 1.5*scale, 0.05*scale],
        [-0.3*scale, 0.9*scale, 0.05*scale],
        [-0.5*scale, 0.9*scale, 0.05*scale],
        
        # Right arm
        [0.3*scale, 1.5*scale, -0.05*scale],
        [0.5*scale, 1.5*scale, -0.05*scale],
        [0.5*scale, 0.9*scale, -0.05*scale],
        [0.3*scale, 0.9*scale, -0.05*scale],
        [0.3*scale, 1.5*scale, 0.05*scale],
        [0.5*scale, 1.5*scale, 0.05*scale],
        [0.5*scale, 0.9*scale, 0.05*scale],
        [0.3*scale, 0.9*scale, 0.05*scale],
        
        # Left leg
        [-0.15*scale, 0, -0.1*scale],
        [0.05*scale, 0, -0.1*scale],
        [0.05*scale, 0.9*scale, -0.1*scale],
        [-0.15*scale, 0.9*scale, -0.1*scale],
        [-0.15*scale, 0, 0.1*scale],
        [0.05*scale, 0, 0.1*scale],
        [0.05*scale, 0.9*scale, 0.1*scale],
        [-0.15*scale, 0.9*scale, 0.1*scale],
        
        # Right leg
        [-0.05*scale, 0, -0.1*scale],
        [0.15*scale, 0, -0.1*scale],
        [0.15*scale, 0.9*scale, -0.1*scale],
        [-0.05*scale, 0.9*scale, -0.1*scale],
        [-0.05*scale, 0, 0.1*scale],
        [0.15*scale, 0, 0.1*scale],
        [0.15*scale, 0.9*scale, 0.1*scale],
        [-0.05*scale, 0.9*scale, 0.1*scale],
    ]
    
    # Define faces (triangles)
    faces = []
    
    # Helper function to add cube faces
    def add_cube_faces(start_idx):
        faces.extend([
            # Front
            [start_idx, start_idx+1, start_idx+2],
            [start_idx, start_idx+2, start_idx+3],
            # Back
            [start_idx+4, start_idx+6, start_idx+5],
            [start_idx+4, start_idx+7, start_idx+6],
            # Top
            [start_idx+2, start_idx+6, start_idx+7],
            [start_idx+2, start_idx+7, start_idx+3],
            # Bottom
            [start_idx, start_idx+5, start_idx+4],
            [start_idx, start_idx+1, start_idx+5],
            # Left
            [start_idx+4, start_idx+7, start_idx+3],
            [start_idx+4, start_idx+3, start_idx+0],
            # Right
            [start_idx+1, start_idx+6, start_idx+5],
            [start_idx+1, start_idx+2, start_idx+6],
        ])
    
    # Add faces for each body part
    add_cube_faces(0)   # Head
    add_cube_faces(8)   # Torso
    add_cube_faces(16)  # Left arm
    add_cube_faces(24)  # Right arm
    add_cube_faces(32)  # Left leg
    add_cube_faces(40)  # Right leg
    
    # Flatten vertices and faces
    vertices_flat = []
    for v in vertices:
        vertices_flat.extend(v)
    
    faces_flat = []
    for f in faces:
        faces_flat.extend(f)
    
    # Create minimal glTF structure
    gltf = {
        "asset": {"version": "2.0"},
        "scene": 0,
        "scenes": [{"nodes": [0]}],
        "nodes": [
            {
                "mesh": 0,
                "name": "Humanoid"
            }
        ],
        "meshes": [
            {
                "primitives": [
                    {
                        "attributes": {"POSITION": 0},
                        "indices": 1,
                        "material": 0
                    }
                ]
            }
        ],
        "materials": [
            {
                "pbrMetallicRoughness": {
                    "baseColorFactor": [0.5, 0.5, 0.5, 1.0],
                    "metallicFactor": 0.3,
                    "roughnessFactor": 0.7
                }
            }
        ],
        "accessors": [
            {
                "bufferView": 0,
                "componentType": 5126,  # FLOAT
                "count": len(vertices),
                "type": "VEC3",
                "min": [-0.5*scale, 0, -0.15*scale],
                "max": [0.5*scale, 2.1*scale, 0.15*scale]
            },
            {
                "bufferView": 1,
                "componentType": 5125,  # UNSIGNED_INT
                "count": len(faces_flat),
                "type": "SCALAR"
            }
        ],
        "bufferViews": [
            {
                "buffer": 0,
                "byteOffset": 0,
                "byteStride": 12,
                "target": 34962
            },
            {
                "buffer": 0,
                "byteOffset": len(vertices_flat) * 4,
                "byteLength": len(faces_flat) * 4,
                "target": 34963
            }
        ],
        "buffers": [
            {
                "byteLength": len(vertices_flat) * 4 + len(faces_flat) * 4
            }
        ]
    }
    
    # Create binary buffer
    buffer_data = bytearray()
    
    # Add vertices as floats
    for v in vertices_flat:
        buffer_data.extend(struct.pack('<f', v))
    
    # Add faces as unsigned ints
    for f in faces_flat:
        buffer_data.extend(struct.pack('<I', f))
    
    # Encode JSON
    json_str = json.dumps(gltf)
    json_bytes = json_str.encode('utf-8')
    
    # Pad JSON to 4-byte boundary
    json_padding = (4 - len(json_bytes) % 4) % 4
    json_bytes += b' ' * json_padding
    
    # Create GLB file
    glb_data = bytearray()
    
    # Header
    glb_data.extend(b'glTF')  # Magic
    glb_data.extend(struct.pack('<I', 2))  # Version
    glb_data.extend(struct.pack('<I', 28 + len(json_bytes) + len(buffer_data)))  # File size
    
    # JSON chunk
    glb_data.extend(struct.pack('<I', len(json_bytes)))  # Chunk size
    glb_data.extend(b'JSON')  # Chunk type
    glb_data.extend(json_bytes)
    
    # Binary chunk
    glb_data.extend(struct.pack('<I', len(buffer_data)))  # Chunk size
    glb_data.extend(b'BIN\x00')  # Chunk type
    glb_data.extend(buffer_data)
    
    # Write file
    filepath = output_dir / filename
    with open(filepath, 'wb') as f:
        f.write(glb_data)
    
    print(f"✅ Created: {filename} ({len(glb_data)} bytes)")
    return filepath


def main():
    print("🤖 Generating humanoid models...\n")
    
    try:
        # Create models
        models = [
            ('SimpleHumanoid.glb', 1.0),
            ('TallHumanoid.glb', 1.2),
            ('CompactHumanoid.glb', 0.8),
        ]
        
        for filename, scale in models:
            create_simple_humanoid_glb(filename, scale)
        
        print(f"\n✨ Model generation complete!")
        print(f"📂 Models saved to: {output_dir}")
        print("\n💡 Next steps:")
        print("1. Import models into 3D Poser app")
        print("2. Test bone selection and rigging")
        print("3. Create and save poses")
        print("4. Record animations with timeline")
        
    except Exception as e:
        print(f"❌ Error: {e}")
        import traceback
        traceback.print_exc()
        exit(1)


if __name__ == '__main__':
    main()
